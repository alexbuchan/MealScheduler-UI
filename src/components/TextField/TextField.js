import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  label: PropTypes.string,              // The name on the label that renders in the UI. /Optional
  name: PropTypes.string.isRequired,    // The name of the field
  type: PropTypes.string,               // The type of input for the field
  value: PropTypes.string.isRequired,   // The value of the input in the field
  placeholder: PropTypes.string,         // Optional placeholder for textfield
  onChange: PropTypes.func.isRequired,  // The onChange function for updating the field
  validationField: PropTypes.object     // Validation object returned from the Validator instance.
};

const TextField = ({ label, name, type, value, placeholder, onChange, validationField }) => {
  const inputFieldClass = () => {
    return (!validationField.isInvalid) ? "" : "input-field-error";
  }

  const withValidation = () => {
    return (
      <div className='form-element'>
        <label>{ label }</label>
        <input
          data-test='input-with-validation'
          className={ inputFieldClass() }
          name={ name }
          type={ type }
          value={ value }
          placeholder={ placeholder }
          onChange={ onChange }
        />
        <span className="help-block">{ validationField.message }</span>
      </div>
    );
  }

  const noValidation = () => {
    return (
      <div className='form-element'>
        <label>{ label }</label>
        <input
          name={ name }
          type={ type }
          value={ value }
          placeholder={ placeholder }
          onChange={ onChange }
        />
      </div>
    );
  }

  if (validationField) {
    return withValidation();
  }

  return noValidation();
}

TextField.defaultProps = {
  type: 'text'
}

TextField.propTypes = propTypes;
export default TextField;