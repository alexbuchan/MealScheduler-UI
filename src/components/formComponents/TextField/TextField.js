import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  label: PropTypes.string,              // The name on the label that renders in the UI. /Optional
  name: PropTypes.string.isRequired,    // The name of the field
  type: PropTypes.string,               // The type of input for the field
  value: PropTypes.string.isRequired,   // The value of the input in the field
  width: PropTypes.number,              // The width property for input
  height: PropTypes.number,             // The height property for input
  disabled: PropTypes.bool,             // Disables the text field
  placeholder: PropTypes.string,        // Optional placeholder for textfield
  onChange: PropTypes.func.isRequired,  // The onChange function for updating the field
  validationField: PropTypes.object,    // Validation object returned from the Validator instance.
  isRequired: PropTypes.bool,           // When validation from a form is turned on, this marks the field as required
  inline: PropTypes.bool                // Places the label inline with the textfield 
};

const TextField = ({ label, name, type, value, width, height, disabled, placeholder, onChange, validationField, inline }) => {
  const inlineLabel = () => {
    return (inline) ? "text-field-inline" : "text-field-component";
  }

  const inputFieldClass = () => {
    return (!validationField.isInvalid) ? "" : "input-field-error";
  }

  const isRequiredLabel = () => {
    if (validationField.isRequired) {
      return <span data-test='text-field-is-required' className='text-field-is-required'>*</span>;
    }

    return null;
  }

  const withValidation = () => {
    return (
      <div className={ inlineLabel() }>
        <div>
          <label>{ label }</label>
          { isRequiredLabel() }
        </div>

        <input
          data-test='input-with-validation'
          className={ `${inputFieldClass()} text-field-input` }
          name={ name }
          style={ { width, height } }
          type={ type }
          value={ value }
          disabled={ disabled }
          placeholder={ placeholder }
          onChange={ onChange }
        />

        <span className="help-block">{ validationField.message }</span>
      </div>
    );
  }

  const noValidation = () => {
    return (
      <div className={ inlineLabel() }>
        <label className='text-field-label'>{ label }</label>
        <input
          name={ name }
          className='text-field-input'
          style={ { width, height } }
          type={ type }
          value={ value }
          disabled={ disabled }
          placeholder={ placeholder }
          onChange={ onChange }
        />
      </div>
    );
  }

  if (validationField) {
    return withValidation();
  }

  return noValidation();
}

TextField.defaultProps = {
  type: 'text',
  inline: false
}

TextField.propTypes = propTypes;
export default TextField;