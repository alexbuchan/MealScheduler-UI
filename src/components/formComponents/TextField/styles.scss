.text-field-component {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.text-field-inline {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.text-field-is-required { 
  flex-direction: row;
  margin-left: 3px;
  font-size: 17px;
  color: red;
}

.text-field-label {
  margin: 0 10px 0 0;
}

.text-field-input {
  height: 41px;
  width: 200px;
  border: 1px solid lightgray;
  border-radius: 4px;
}

.text-field-input::-webkit-calendar-picker-indicator {
  opacity: 0.75;
}

.text-field-input::-webkit-datetime-edit-fields-wrapper {
  opacity: 0.3;
  margin-left: 13px;
}