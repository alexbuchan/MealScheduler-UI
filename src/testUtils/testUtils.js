import React from 'react';
import { shallow } from 'enzyme';

export const componentSetup = (Component, props = {}, options = {}) => {
  const instance = shallow(<Component { ...props } />, options);
  return instance;
}

export const findByTestAttribute = (instance, attribute) => {
  const wrapper = instance.find(`[data-test="${attribute}"]`);
  return wrapper;
}