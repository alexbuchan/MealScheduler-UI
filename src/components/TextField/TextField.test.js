import TextField from './TextField';
import { componentSetup, findByTestAttribute } from '../../testUtils/testUtils';

describe('TextField', () => {
  describe('When validation is disabled', () => {
    describe('When no label is specified', () => {
      let instance;

      beforeEach(() => {
        const props = {
          name: 'name',
          value: 'value',
          onChange: () => {}
        };

        instance = componentSetup(TextField, props);
      });
      
      it('Should not render a label', () => {
        expect(instance).toMatchSnapshot();
      });
    });

    describe('When no type is specified', () => {
      let instance;

      beforeEach(() => {
        const props = {
          label: 'label',
          name: 'name',
          value: 'value',
          onChange: () => {}
        };

        instance = componentSetup(TextField, props);
      });
      
      it('Should render as type text', () => {
        expect(instance).toMatchSnapshot();
      });
    });

    describe('When all props are specified expect validationField', () => {
      let instance;

      beforeEach(() => {
        const props = {
          label: 'label',
          name: 'name',
          type: 'text',
          value: 'value',
          onChange: () => {}
        };

        instance = componentSetup(TextField, props);
      });
      
      it('Should render correctly', () => {
        expect(instance).toMatchSnapshot();
      });
    });
  });

  describe('When validation is enabled', () => {
    describe('When input is valid', () => {
      let instance;

      beforeEach(() => {
        const props = {
          label: 'label',
          name: 'name',
          type: 'text',
          value: 'valid value',
          onChange: () => {},
          validationField: { isInvalid: false, message: '' }
        };

        instance = componentSetup(TextField, props);
      });

      it('should not render validation error message', () => {
        expect(instance).toMatchSnapshot();
      });

      it('input className should be an empty string', () => {
        expect(findByTestAttribute(instance, 'input-with-validation').hasClass('')).toBe(true);
      });
    });

    describe('When input is NOT valid', () => {
      let instance;

      beforeEach(() => {
        const props = {
          label: 'label',
          name: 'name',
          type: 'text',
          value: 'invalid value',
          onChange: () => {},
          validationField: { isInvalid: true, message: 'Value must be valid' }
        };

        instance = componentSetup(TextField, props);
      });

      it('should render validation error message', () => {
        expect(instance).toMatchSnapshot();
      });

      it('input className should be "input-field-error"', () => {
        expect(findByTestAttribute(instance, 'input-with-validation').hasClass('input-field-error')).toBe(true);
      });
    });
  });
});