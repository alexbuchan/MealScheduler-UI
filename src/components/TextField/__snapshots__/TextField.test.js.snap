// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField When validation is disabled When all props are specified expect validationField Should render correctly 1`] = `
<div
  className="form-element"
>
  <label>
    label
  </label>
  <input
    name="name"
    onChange={[Function]}
    type="text"
    value="value"
  />
</div>
`;

exports[`TextField When validation is disabled When no label is specified Should not render a label 1`] = `
<div
  className="form-element"
>
  <label />
  <input
    name="name"
    onChange={[Function]}
    type="text"
    value="value"
  />
</div>
`;

exports[`TextField When validation is disabled When no type is specified Should render as type text 1`] = `
<div
  className="form-element"
>
  <label>
    label
  </label>
  <input
    name="name"
    onChange={[Function]}
    type="text"
    value="value"
  />
</div>
`;

exports[`TextField When validation is enabled When input is NOT valid should render validation error message 1`] = `
<div
  className="form-element"
>
  <label>
    label
  </label>
  <input
    className="input-field-error"
    data-test="input-with-validation"
    name="name"
    onChange={[Function]}
    type="text"
    value="invalid value"
  />
  <span
    className="help-block"
  >
    Value must be valid
  </span>
</div>
`;

exports[`TextField When validation is enabled When input is valid should not render validation error message 1`] = `
<div
  className="form-element"
>
  <label>
    label
  </label>
  <input
    className=""
    data-test="input-with-validation"
    name="name"
    onChange={[Function]}
    type="text"
    value="valid value"
  />
  <span
    className="help-block"
  />
</div>
`;
